src/audio/template_comp/template_comp.c

```diff
@@ -0,0 +1,309 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2025 Intel Corporation.
+
+#include <sof/audio/module_adapter/module/generic.h>
+#include <rtos/init.h>
+#include "template_comp.h"
+
+LOG_MODULE_REGISTER(template_comp, CONFIG_SOF_LOG_LEVEL);
+
+/* Use e.g. command uuidgen from package uuid-runtime, add it to
+ * uuid-registry.txt in SOF top level.
+ */
+SOF_DEFINE_REG_UUID(template_comp);
+
+DECLARE_TR_CTX(template_comp_tr, SOF_UUID(template_comp_uuid), LOG_LEVEL_INFO);
+
+static int template_comp_init(struct processing_module *mod)
+{
+	struct module_data *md = &mod->priv;
+	struct comp_dev *dev = mod->dev;
+	struct template_comp_comp_data *cd;
+
+	comp_info(dev, "template_comp_init()");
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd)
+		return -ENOMEM;
+
+	md->private = cd;
+	return 0;
+}
+
+__cold static int template_comp_free(struct processing_module *mod)
+{
+	struct template_comp_comp_data *cd = module_get_private_data(mod);
+
+	assert_can_be_cold();
+
+	comp_info(mod->dev, "template_comp_free()");
+	rfree(cd);
+	return 0;
+}
+
+#if CONFIG_FORMAT_S16LE
+static void template_comp_s16(const struct processing_module *mod,
+			      const struct audio_stream *source,
+			      struct audio_stream *sink,
+			      uint32_t frames)
+{
+	struct template_comp_comp_data *cd = module_get_private_data(mod);
+	int16_t *x = audio_stream_get_rptr(source);
+	int16_t *y = audio_stream_get_wptr(sink);
+	int nbuf;
+	int nfrm;
+	int ch;
+	int i;
+
+	while (frames) {
+		nbuf = audio_stream_frames_without_wrap(source, x);
+		nfrm = MIN(frames, nbuf);
+		nbuf = audio_stream_frames_without_wrap(sink, y);
+		nfrm = MIN(nfrm, nbuf);
+		for (i = 0; i < nfrm; i++) {
+			for (ch = 0; ch < cd->channels; ch++) {
+				*y = *(x + cd->channels_order[ch]);
+				y++;
+			}
+			x += cd->channels;
+		}
+		frames -= nfrm;
+		x = audio_stream_wrap(source, x);
+		y = audio_stream_wrap(sink, y);
+	}
+}
+#endif /* CONFIG_FORMAT_S16LE */
+
+#if CONFIG_FORMAT_S32LE
+/* Same function works for s24 and s32 since
+ * the samples values are not modified.
+ */
+static void template_comp_s32(const struct processing_module *mod,
+			      const struct audio_stream *source,
+			      struct audio_stream *sink,
+			      uint32_t frames)
+{
+	struct template_comp_comp_data *cd = module_get_private_data(mod);
+	int32_t *x = audio_stream_get_rptr(source);
+	int32_t *y = audio_stream_get_wptr(sink);
+	int nbuf;
+	int nfrm;
+	int ch;
+	int i;
+
+	while (frames) {
+		nbuf = audio_stream_frames_without_wrap(source, x);
+		nfrm = MIN(frames, nbuf);
+		nbuf = audio_stream_frames_without_wrap(sink, y);
+		nfrm = MIN(nfrm, nbuf);
+		for (i = 0; i < nfrm; i++) {
+			for (ch = 0; ch < cd->channels; ch++) {
+				*y = *(x + cd->channels_order[ch]);
+				y++;
+			}
+			x += cd->channels;
```

Comment:
maybe in a follow-up - should we add template-generic.c and template-hifi5.c with different versions of these functions? Then also a template_comp.h header becomes justified, otherwise it isn't really needed?

Reply:
I need to complete this task quickly for coming power-on and move on to other critical things, so I'd like to postpone the HiFi5 version. But sure it makes sense, and shows how to use the Xtensa version helper macros.

Reply:
Also per my own experience the xt-clang compiler for HiFi5 is good, it's hard to beat the performance of good C code with hand written intrinsics. Intrinsics are good for saturated math with rounding that is not efficient with C but otherwise the gain can be low.

Reply:
@singalsu I think we can move this functions now into a file name `template-generic.c`. It doesn't pollute the `template_comp.c` code and makes everything easier to read.